File: apps/api/app/core/admin_email_allowlist.py
"""
ADMIN_EMAILS env: comma-separated; entries compared via normalize_email_for_identity.

Keep allowlist parsing rules aligned with: apps/web/lib/admin-email-allowlist.ts
"""

from __future__ import annotations

import os
from functools import lru_cache

from app.core.email_identity import normalize_email_for_identity


@lru_cache(maxsize=1)
def _normalized_admin_email_set() -> frozenset[str]:
    raw = (os.environ.get("ADMIN_EMAILS") or "").strip()
    out: list[str] = []
    for part in raw.split(","):
        s = part.strip()
        if not s:
            continue
        norm = normalize_email_for_identity(s)
        if norm:
            out.append(norm)
    return frozenset(out)


def clear_admin_email_cache() -> None:
    """For tests or reload scenarios."""
    _normalized_admin_email_set.cache_clear()


def email_is_admin_allowlisted(email: str | None) -> bool:
    if not email or not isinstance(email, str):
        return False
    norm = normalize_email_for_identity(email)
    if not norm:
        return False
    return norm in _normalized_admin_email_set()

File: apps/web/lib/admin-email-allowlist.ts
/**
 * ADMIN_EMAILS: comma-separated; entries and candidates use normalizeEmailForIdentity.
 * Aligned with: apps/api/app/core/admin_email_allowlist.py
 */

import { normalizeEmailForIdentity } from "@/lib/email-identity";

export function parseAdminEmailAllowlist(envValue: string | undefined): Set<string> {
  const raw = (envValue ?? "").trim();
  if (!raw) return new Set();
  const out = new Set<string>();
  for (const part of raw.split(",")) {
    const s = part.trim();
    if (!s) continue;
    const norm = normalizeEmailForIdentity(s);
    if (norm) out.add(norm);
  }
  return out;
}

/** Server-only: reads process.env.ADMIN_EMAILS (no NEXT_PUBLIC_). */
export function isAdminEmailAllowlisted(email: string | null | undefined): boolean {
  if (!email || typeof email !== "string") return false;
  const norm = normalizeEmailForIdentity(email);
  if (!norm) return false;
  const allow = parseAdminEmailAllowlist(process.env.ADMIN_EMAILS);
  return allow.has(norm);
}
